plone/formwidget/namedfile/utils.py

```diff
@@ -0,0 +1,46 @@
+# -*- coding: utf-8 -*-
+from datetime import datetime
+from datetime import timedelta
+from plone.formwidget.namedfile.interfaces import IFileUploadMap
+from zope.annotation.interfaces import IAnnotations
+from zope.component import adapter
+from zope.interface import implementer
+from zope.interface import Interface
+from ZPublisher.HTTPRequest import FileUpload
+from BTrees.IOBTree import IOBTree
+from random import randint
+
+FILE_UPLOAD_MAP_KEY = 'file_upload_map'
+FILE_UPLOAD_EXPIRATION_TIME = 30*60  # seconds
+
+
+def is_file_upload(item):
+    """Check if ``item`` is a file upload.
+    """
+    return isinstance(item, FileUpload)
+
+
+@implementer(IFileUploadMap)
+@adapter(Interface)
+class FileUploadMap(object):
+    """Temporary storage adapter for file uploads.
+    To be used to not need to re-upload files after form submission errors.
+    """
+
+    def __init__(self, context):
+        self.context = context
+
+    @property
+    def upload_map(self):
+        annotations = IAnnotations(self.context)
+        upload_map = annotations.get(FILE_UPLOAD_MAP_KEY, IOBTree())
+
+        for key, val in upload_map.items():
+            if val.get('dt', datetime(2000, 2, 2)) < (
+                datetime.now() - timedelta(
+                    seconds=FILE_UPLOAD_EXPIRATION_TIME
+                )
+            ) and randint(0, 5) == 0:  # Avoid conflict errors by deleting only every fifth time  # noqa
```

Comment:
As requested, but every fifth time. The Upload map is only called when a upload is done, not on every request.

Reply:
well, should be ok.

Reply:
Minor point: `datetime(2000, 2, 2)` is instantiated for every item in the map, unless maybe Python sees it as static and compiles it once, but I guess not. So I suggest to define this in the top of the file as `FALLBACK_DATE` or similar.